vllm/attention/ops/blocksparse_attention/utils.py

```diff
@@ -0,0 +1,223 @@
+        num_blocks = triton.cdiv(max_seqlen, block_size)
+        q_pos = torch.arange(num_blocks)[:, None]
+        k_pos = torch.arange(num_blocks)[None]
+        mask_vert_strided = (torch.arange(num_blocks) + 1) % vert_stride == 0
+        block_mask_dense = (((q_pos >= k_pos)
+                             & ((q_pos - k_pos < local_blocks)
+                                | mask_vert_strided)).to(device).to(dtype))
+        num_blocks_q = triton.cdiv(q_len, block_size)
+        block_mask_dense_output = (
+            block_mask_dense[-num_blocks_q:].contiguous().to_sparse_csr())
+    if return_dense:
+        mask_dense = torch.kron(
+            block_mask_dense,
+            block_mask_dense.new_ones((block_size, block_size)),
+        )
+        causal_mask = torch.tril(torch.ones(
+            max_seqlen, max_seqlen)).type_as(mask_dense)[-q_len:]
+        mask_dense = mask_dense[-q_len:, :max_seqlen] * causal_mask
+        return (
+            (
+                block_mask_dense_output.crow_indices(),
+                block_mask_dense_output.col_indices(),
+            ),
+            block_mask_dense,
+            mask_dense,
+        )
+    else:
+        return (
+            (
+                block_mask_dense_output.crow_indices(),
+                block_mask_dense_output.col_indices(),
+            ),
+            block_mask_dense,
+            None,
+        )
+
+
+def binary_mask_to_bias(mask_dense):
+    mask_dense = 1 - mask_dense
+    mask_dense.masked_fill_(mask_dense.bool(), -torch.inf)
+    return mask_dense
+
+
+def get_head_sliding_step(n_heads, vert_stride, homo_head=False):
+    # if vert_stride <= n_heads, rotating the heads
+    if homo_head:
+        return 0
+    return max(1, int(vert_stride / n_heads))
+
+
+@lru_cache
+def get_sparse_attn_mask(
+    n_heads,
+    q_len,
+    max_seqlen,
+    dtype,
+    device,
+    block_size=128,
+    local_blocks=4,
+    vert_stride=4,
+    homo_head=True,
+    return_dense=False,
+    dense_mask_type="binary",
+):
+    """
+    :param dense_mask_type: "binary" (0 for skip token, 1 for others)
+        or "bias" (-inf for skip token, 0 or others)
+    :return: a tuple of 3:
+        - tuple of crow_indices, col_indices representation 
+            of CSR format.
+        - block dense mask
+        - all token dense mask (be aware that it can be OOM if it 
+            is too big) if `return_dense==True`, otherwise, None
+    """
+    assert dense_mask_type in ("binary", "bias")
+    if homo_head:
+        with torch.no_grad():
+            (crow, col), block_mask_dense, mask_dense = (
+                _get_sparse_attn_mask_homo_head(
+                    q_len,
+                    max_seqlen,
+                    dtype,
+                    device,
+                    block_size,
+                    local_blocks,
+                    vert_stride,
+                    return_dense,
+                ))
+            crow = crow[None].expand(n_heads, crow.shape[0])
```

Comment:
I've never seen `[None]` indexing used, what is this achieving?

Reply:
`tensor[None]` is same as `tensor.unsqueeze(0)`, expanding a trivial dimension in both `torch` and `numpy`, pretty standard operation.